cmake/check_linker_flag.cmake

```diff
@@ -0,0 +1,53 @@
+include(CheckCXXSourceCompiles)
+# We need some extra FAIL_REGEX patterns
+# Note that CHECK_C_SOURCE_COMPILES is a misnomer, it will also link.
+SET(fail_patterns
+    FAIL_REGEX "argument unused during compilation"
+    FAIL_REGEX "unsupported .*option"
+    FAIL_REGEX "unknown .*option"
+    FAIL_REGEX "unrecognized .*option"
+    FAIL_REGEX "ignoring unknown option"
+    FAIL_REGEX "warning:.*ignored"
+    FAIL_REGEX "warning:.*is valid for.*but not for"
+    FAIL_REGEX "warning:.*redefined"
+    FAIL_REGEX "[Ww]arning: [Oo]ption"
+    )
+#The regex patterns above are not localized, thus LANG=C
+SET(ENV{LANG} C)
+
+MACRO (MY_CHECK_MODULE_LINKER_FLAG flag)
+  STRING(REGEX REPLACE "[-,= +]" "_" result "HAVE_MODULE_${flag}")
+  SET(SAVE_CMAKE_REQUIRED_LINK_OPTIONS "${CMAKE_REQUIRED_LINK_OPTIONS}")
+  SET(CMAKE_REQUIRED_LINK_OPTIONS ${CMAKE_REQUIRED_LINK_OPTIONS} ${flag})
+  CHECK_CXX_SOURCE_COMPILES("int main(void) { return 0; }" ${result}
+    ${fail_patterns})
+  SET(CMAKE_REQUIRED_LINK_OPTIONS "${SAVE_CMAKE_REQUIRED_LINK_OPTIONS}")
+ENDMACRO()
+
+FUNCTION(MY_CHECK_AND_SET_LINKER_FLAG flag_to_set)
+  # At the moment this is gcc-only.
+  # Let's avoid expensive compiler tests on Windows:
+  IF(WIN32)
+    RETURN()
+  ENDIF()
+  STRING(REGEX REPLACE "^-Wno-" "-W" flag_to_check ${flag_to_set})
+  #MY_CHECK_SHARED_LINKER_FLAG(${flag_to_check})
+  MY_CHECK_MODULE_LINKER_FLAG(${flag_to_check})
+  STRING(REGEX REPLACE "[-,= +]" "_" result "${flag_to_check}")
+  # TODO - SET(HAVE_SHARED..) is temporary only. remove once
+  # MY_CHECK_SHARED_LINKER_FLAG perfoms independent check.
+  SET(HAVE_SHARED_${result} HAVE_MODULE_${result})
+  FOREACH(linktype SHARED MODULE)
```

Comment:
What about `CMAKE_EXE_LINKER_FLAGS`? Do you think it's not needed here?

Reply:
Good call, its needed.